### PR TITLE
Refs #16614 -- Made QuerySet.iterator() use server-side cursors on PostgreSQL.

### PR DESCRIPTION
**Pull request by francoisfreitag**
Refs [ticket 16614](https://code.djangoproject.com/ticket/16614)

Based on Josh Smeaton idea of implementing server side cursors in PostgreSQL from the iterator method and previous work by Anssi Kääriäinen and Kevin Turner.

If this PR is accepted, I will be pleased to add some lines in the release note, but I think some discussion should occur first. In particular:

- ~~`.iterator()` is called in [`_fetch_all()`](https://github.com/django/django/blob/master/django/db/models/query.py#L1077), making server-side cursors widely used.~~ Resolved, thanks to @charettes.
- Currently, query results are loaded into memory by `psycopg`. This PR transfers the resource usage from the Python worker to the database. An opt-out method seems important here, but I don't have any idea beyond (another) setting.
- In autocommit mode, the named cursors are declared as `WITH HOLD`, which ties database resources until the cursor is deleted, this could significantly increase resource consumption.

Note:
I used separate commits to change existing tests. The commit message should help reviewing the changes.
I also kept the initial patch from Anssi and Kevin untouched.

### CODE REVIEW

File: django/db/models/query.py [hunk 1]
```diff
@@ -321,6 +321,7 @@ def iterator(self):
         An iterator over the results from applying this QuerySet to the
         database.
         """
+        self.query.chunked_fetch = True
```
Comment (charettes) [inline django/db/models/query.py]: We might want to avoid storing this flag in `self.query` as iterating over this queryset (invoking `self.__iter__()`) would also use `chunked_fetch`. As you've mentioned this doesn't change anything now since `self._fetch_all()` calls `iterator()` under the hood but I don't think we want to use server side cursors for every query.

I suggest me pass `chunked_fetch` as a parameter to `self._iterable_class()` and make `_fetch_all` uses `list(self._iterable_class(self))` instead of calling `iterator()`.

Passing the _chunked_ parameter on iterable instantiation could also us to define a `__del__` method on them to delete the named cursor in cases where the iterator is not completely exhausted but falls out of scope in autocommit mode.

Comment (francoisfreitag) [inline django/db/models/query.py]: I am not sure I understand your last point. When the reference to the `QuerySet` falls out of scope, the last reference to the cursor is [finally closed](https://github.com/django/django/blob/master/django/db/models/sql/compiler.py#L1224-L1225), even when the iterator is not completely exhausted. Or maybe I have missed something?

I did the following experiment, it would seem to indicate that the finally clause is called even when the iterator has not been fully consumed.
```python
class Cursor():
    def __del__(self):
        print('Cursor %r deleted' % self)

def cursor_iter():
    try:
        cur = Cursor()
        while True:
            yield cur
    finally:
        print('End of generator, deleting cursor')

def user():
    c = cursor_iter()
    for _ in range(3):
        next(c)
    # c is not totally consumed and falls out of scope

user()
```
prints:
```
End of generator, deleting cursor
Cursor <__main__.Cursor instance at 0x7fe840846758> deleted
```

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -321,6 +322,7 @@ def clone(self, klass=None, memo=None, **kwargs):
             obj.used_aliases = set()
         obj.filter_is_sticky = False
         obj.subquery = self.subquery
+        obj.chunked_fetch = self.chunked_fetch
```
Comment (charettes) [inline django/db/models/sql/query.py]: Both of these are not required anymore.

**Comment by charettes** [comment]
In this case the `__del__` part might not be required. I didn't know
generators `finally` were called when they fell out of scope and that we
had this code in place.

2016-11-22 13:12 GMT-05:00 François Freitag <notifications@github.com>:

File: django/db/models/sql/compiler.py [hunk 3]
```diff
@@ -852,11 +856,12 @@ def execute_sql(self, result_type=MULTI):
...
+        if not self.connection.features.can_use_chunked_reads and not chunked_fetch:
             try:
                 # If we are using non-chunked reads, we return the same data
                 # structure as normally, but ensure it is all read into memory
-                # before going any further.
+                # before going any further. If the user requests chunked_fetch
+                # we will honor that.
                 return list(result)
             finally:
                 # done with the cursor
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: Not related to this patch but I wonder if the `cursor.close()` below should also be wrapped?

File: django/db/models/sql/compiler.py [hunk 4]
```diff
@@ -852,11 +856,12 @@ def execute_sql(self, result_type=MULTI):
             cursor, self.connection.features.empty_fetchmany_value,
             self.col_count
         )
-        if not self.connection.features.can_use_chunked_reads:
+        if not self.connection.features.can_use_chunked_reads and not chunked_fetch:
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: Might want to move the `not chunked_fetch` local variable checks before performing a double attribute lookup  with `self.connection.features.can_use_chunked_reads`.

File: django/db/models/sql/compiler.py (hunk 3 above)
Comment (francoisfreitag) [inline django/db/models/sql/compiler.py]: Nice catch.

**Comment by jarshwah** [comment]
Awesome work on this, thanks for writing it! I haven't gone over the patch in detail yet, but I am somewhat concerned about the naming of the cursors. Mainly, what happens when a separate thread or a separate process generate cursors with the same name at the same time?

I'm assuming the name of the cursor needs to be unique to the database at the time it's executed, but I've been unable to find great documentation that explains this. It'd be nice to see some analysis done here, especially if we can either catch the error and report it more nicely than Postgres, or consider switching to a different naming scheme.

**Comment by francoisfreitag** [comment]
I've tested the naming conflicts quickly while developing it, directly in the PostgreSQL shell. The cursor name is only valid inside a connection, so it's possible for two threads/processes to use the same cursor name at a given time, as long as the connection is not shared.

My experiment was the following, on any PostgreSQL instance:
```sql
-- $ psql django
django=# DECLARE cursor_1 CURSOR WITH HOLD FOR SELECT * FROM pg_tables;
DECLARE CURSOR
django=# SELECT * FROM pg_cursors;
   name   |                           statement                            | is_holdable | is_binary | is_scrollable |         creation_time
----------+----------------------------------------------------------------+-------------+-----------+---------------+-------------------------------
 cursor_1 | DECLARE cursor_1 CURSOR WITH HOLD FOR SELECT * FROM pg_tables; | t           | f         | f             | 2016-12-19 22:05:18.980156-08
(1 row)
-- Leave the cursor open, and pop-up a second shell
```
```sql
-- $ psql django
-- The first shell still holds cursor_1
django=# SELECT * FROM pg_cursors;
 name | statement | is_holdable | is_binary | is_scrollable | creation_time
------+-----------+-------------+-----------+---------------+---------------
(0 rows)
-- The cursor from the other connection is not listed

django=# DECLARE cursor_1 CURSOR WITH HOLD FOR SELECT * FROM pg_tables;
DECLARE CURSOR
-- It is possible to create a cursor with the same name

django=# SELECT * FROM pg_cursors;
   name   |                           statement                            | is_holdable | is_binary | is_scrollable |         creation_time
----------+----------------------------------------------------------------+-------------+-----------+---------------+-------------------------------
 cursor_1 | DECLARE cursor_1 CURSOR WITH HOLD FOR SELECT * FROM pg_tables; | t           | f         | f             | 2016-12-19 22:09:37.979431-08

-- For compleness, verify that both connections are active:
 datid | datname |  pid  | usesysid | usename  | application_name | client_addr | client_hostname | client_port |         backend_start         |          xact_start           |          query_start          |         state_change          | wait_event_type | wait_event | state  | backend_xid | backend_xmin |              query
-------+---------+-------+----------+----------+------------------+-------------+-----------------+-------------+-------------------------------+-------------------------------+-------------------------------+-------------------------------+-----------------+------------+--------+-------------+--------------+---------------------------------
 16385 | django  | 25066 |    16384 | freitafr | psql             |             |                 |          -1 | 2016-12-19 22:04:14.021095-08 |                               | 2016-12-19 22:06:09.700027-08 | 2016-12-19 22:06:09.700246-08 |                 |            | idle   |             |              | SELECT * FROM pg_cursors;
 16385 | django  | 25600 |    16384 | freitafr | psql             |             |                 |          -1 | 2016-12-19 22:09:04.881431-08 | 2016-12-19 22:11:47.210779-08 | 2016-12-19 22:11:47.210779-08 | 2016-12-19 22:11:47.210781-08 |                 |            | active |             |        45532 | SELECT * FROM pg_stat_activity;
(2 rows)
```

Given [Databases caveats](https://docs.djangoproject.com/en/1.10/ref/databases/#caveats) and a quick glance to  `allow_thread_sharing`, it would seem that two threads will not share the same connection in the default use case. I'll investigate the behavior when the connection is shared between two threads and post the result on the PR in the next couple of days.

File: docs/ref/models/querysets.txt [hunk 5]
```diff
@@ -1982,16 +1982,6 @@ evaluated will force it to evaluate again, repeating the query.
...
 
-.. warning::
-
-    Some Python database drivers like ``psycopg2`` perform caching if using
-    client side cursors (instantiated with ``connection.cursor()`` and what
-    Django's ORM uses). Using ``iterator()`` does not affect caching at the
-    database driver level. To disable this caching, look at `server side
-    cursors`_.
-
-.. _server side cursors: http://initd.org/psycopg/docs/usage.html#server-side-cursors
```
Comment (jarshwah) [inline docs/ref/models/querysets.txt]: I wonder if this warning is worth keeping around, but altering to describe bundled backends that do and do not support server side cursors.

**Comment by jarshwah** [comment]
Can you write up the release notes too please? Once those and the iterator docs are written, it should be very close to merging.

**Comment by francoisfreitag** [comment]
Unsurprisingly, attempting to share a cursor name between two threads using the same connection ends up as a `psycopg2.ProgrammingError: cursor "test" already exists`.

I propose to include the thread `ident` in the cursor name to prevent this conflict from happening. The max length for a cursor name is 63 chars. I think it's safe to include the approx. 15 digits of the thread ident, leaving roughly `10**34` cursor names (63 - 13 (`_django_curs_`) - 16 (`threading.get_ident()` + `_`)).

**Comment by apollo13** [comment]
Since I am currently on #postgresql to get help for my UNION patch, I also asked about server side cursors. Those are the answers I got:
```
<StuckMojo> they're not as expensive in PG as other RDBMSes, but the multiple network round trips would make them slower overall than a plain query

<RhodiumToad> there's the issue of cursor_tuple_fraction to watch out for
<RhodiumToad> by default, when pg plans a cursor query, it assumes that only the first 10% of the result will be fetched
<RhodiumToad> that can get you a fast-start plan, but one that takes excessive time if you actually fetch the whole result
<RhodiumToad> look up the cursor_tuple_fraction setting in the docs
<RhodiumToad> you can (and should) set that to 1 if you're declaring a cursor where you plan to fetch all the rows
```
So maybe we might wanna be conservative and do not enable this by default

**Comment by jarshwah** [comment]
That's some good information to have, but I think the primary use case of using `.iterator()` is for memory constraint issues, not really query performance. The internet is full of people surprised when hitting Out Of Memory errors because they mostly assume that `iterator()` does what this patch is implementing.

We could document the `cursor_tuple_fraction` setting somewhere in our docs, and allow database administrators to choose what the setting should be for their work load. Creating a new function only delays the problem for DBAs and would make the situation less clear for users whether they should be using `iterator()` or some new function that does the same thing except on Postgres/Oracle.

**Comment by charettes** [comment]
+1 to what Josh said.

@Xof do you have any input about making `iterator()` default to server-side cursors on PostgreSQL.

**Comment by apollo13** [comment]
@jarshwah Mhm, I have been bitten by bad performance of server side cursors quite a few times in the past. But I see that psycopg2 fetches 2000 by default, so that should be okayish, would it be possible to pass that `itersize` (http://initd.org/psycopg/docs/usage.html#server-side-cursors) to our `.iterator()`?

**Comment by jarshwah** [comment]
I think being able to pass options to `iterator()` like window size is a good idea, but not something that should hold up this patch. If someone wants to add it before 1.11 that'd be cool though.

**Comment by francoisfreitag** [comment]
I created a [ticket](https://code.djangoproject.com/ticket/27639) to add a batch size option for the `.iterator()`. If it is accepted, I would gladly work on it and try to submit a PR before the release of Django 1.11.

File: docs/ref/models/querysets.txt [hunk 6]
```diff
@@ -1981,15 +1983,19 @@ evaluated will force it to evaluate again, repeating the query.
 Also, use of ``iterator()`` causes previous ``prefetch_related()`` calls to be
 ignored since these two optimizations do not make sense together.
 
-.. warning::
+.. note::
+
+    Some Python database drivers perform caching using client side cursors
+    (instantiated with ``connection.cursor()`` and what Django's ORM uses) and
+    cannot stream results from the database. The following database engines
```
Comment (jarshwah) [inline docs/ref/models/querysets.txt]: I'm not sure this first sentence adds much value. I know it's adapted from the previous documentation, but it reads strange to me, especially since the previous advice (about drivers doing their own caching) seemed to only apply to psycopg2 which is no longer relevant.

How about replacing the first sentence with something like:

> Some Python database drivers will still load the entire result set into memory, but won't cache results after iterating them. The following database engines support streaming results from the database without loading the entire result set into memory (using server-side cursors):

@timgraham what do you think about this wording? As far as I'm concerned the code is fine, but docs might need some light touches, which is why I haven't marked it Ready for Checkin yet.

**Comment by francoisfreitag** [comment]
I removed the commit about the weird behavior in PostgreSQL. It seems that the problem is specific to my machine, since the test suite runs fine on Jenkins.

File: tests/backends/tests.py [hunk 7]
```diff
@@ -1248,3 +1249,49 @@ def create_object():
...
 
         self.assertEqual(Object.objects.count(), 2)
+
+
+@unittest.skipUnless(connection.vendor == 'postgresql', "Test only for PostgreSQL")
+class ServerSideCursorsPostgres(TestCase):
+    cursor_fields = 'name, statement, is_holdable, is_binary, is_scrollable, creation_time'
+    PostgresCursor = namedtuple('PostgresCursor', cursor_fields)
+
+    def setUp(self):
```
Comment (timgraham) [inline tests/backends/tests.py]: setUpTestData?

File: django/db/backends/base/base.py [hunk 8]
```diff
@@ -214,10 +214,25 @@ def ensure_connection(self):
...
+            wrapped_cursor = self.make_debug_cursor(cursor)
+        else:
+            wrapped_cursor = self.make_cursor(cursor)
+        return wrapped_cursor
+
+    def _cursor(self, name=None):
         self.ensure_connection()
         with self.wrap_database_errors:
-            return self.create_cursor()
+            if name:
```
Comment (timgraham) [inline django/db/backends/base/base.py]: Did you consider updating the `def create_cursor` definition on the other backends so if/else isn't needed? This could be mentioned in the "Database backend API" changes in the release notes.

If we want to keep both signatures, something like `return self.create_cursor(*(name,) if name else ())` might be less repetitive. I see this pattern is used a few times.

File: docs/ref/databases.txt [hunk 9]
```diff
@@ -171,6 +171,19 @@ If you need to add a PostgreSQL extension (like ``hstore``, ``postgis``, etc.)
 using a migration, use the
 :class:`~django.contrib.postgres.operations.CreateExtension` operation.
 
+Server side cursors
+-------------------
+
+When using :ref:`queryset_iterator`, Django opens a :ref:`server-side cursor
+<psycopg2:server-side-cursors>`. By default, when PostgreSQL plans a cursor
+query, it assumes that only the first 10% of the results will be fetched. This
+results in a fast-start plan, but can take excessive time if the whole results
```
Comment (timgraham) [inline docs/ref/databases.txt]: I'm not sure what "fast-start" means. Is it a technical term? A Google search was unhelpful. 
"if the whole results" -- what about if more than 10% of results are retrieved?

File: django/test/testcases.py [hunk 10]
```diff
@@ -175,13 +175,15 @@ def setUpClass(cls):
             for alias in connections:
                 connection = connections[alias]
                 connection.cursor = _CursorFailure(cls.__name__, connection.cursor)
+                connection.chunked_cursor = _CursorFailure(cls.__name__, connection.chunked_cursor)
```
Comment (timgraham) [inline django/test/testcases.py]: Is a regression test needed for this change? I don't see any failures after reverting the changes to this file.

File: django/db/models/sql/compiler.py [hunk 11]
```diff
@@ -852,15 +856,17 @@ def execute_sql(self, result_type=MULTI):
...
                 # If we are using non-chunked reads, we return the same data
                 # structure as normally, but ensure it is all read into memory
-                # before going any further.
+                # before going any further. If the user requests chunked_fetch
+                # we will honor that.
                 return list(result)
             finally:
-                # done with the cursor
-                cursor.close()
+                with self.connection.wrap_database_errors:
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: Possible to test this change?

File: tests/prefetch_related/tests.py [hunk 12]
```diff
@@ -252,7 +252,8 @@ def test_m2m_then_reverse_one_to_one_object_ids(self):
             list(Book.objects.prefetch_related('authors__authorwithage'))
 
         sql = queries[-1]['sql']
-        self.assertEqual(sql.count(str(self.author1.id)), 1, sql)
+        author_id_count = sql.count(str(self.author1.id), sql.index('WHERE'))
```
Comment (timgraham) [inline tests/prefetch_related/tests.py]: Maybe you can make this change in a separate commit and add a test assertion something like `assertWhereContains()` and apply it to the other tests also.

File: docs/ref/databases.txt (hunk 9 above)
Comment (francoisfreitag) [inline docs/ref/databases.txt]: I took the "fast-start" term from RoadiumToad in Florian's response. From what I understand, the query planner prefers to start returning results faster rather than taking time to plan the most efficient way to get all results.

I think it would be clearer to say:
By default, when PostgreSQL plans a cursor query, it assumes that only the first 10% of the results will be fetched. The query planner will spend less time planning the query and start returning rows faster, but this could diminish the query performance if more than 10% of the results are retrieved.

File: django/db/models/sql/compiler.py (hunk 11 above)
Comment (francoisfreitag) [inline django/db/models/sql/compiler.py]: I'm not sure if this change is necessary. This finally clause would only get executed with client-side cursors. In most places, the code does not wrap errors on `cursor.close()`, for example:
- https://github.com/francoisfreitag/django/blob/16614/server_side_cursor/django/contrib/gis/db/backends/postgis/introspection.py#L69
- https://github.com/francoisfreitag/django/blob/16614/server_side_cursor/django/contrib/gis/db/backends/spatialite/introspection.py#L60
- https://github.com/francoisfreitag/django/blob/16614/server_side_cursor/django/contrib/gis/db/backends/oracle/introspection.py#L48

I don't see why wrapping `cursor.close()` here would be necessary?

Comment (francoisfreitag) [inline django/db/models/sql/compiler.py]: Should I remove this wrapping before this PR is merged?

Comment (timgraham) [inline django/db/models/sql/compiler.py]: I think I removed it in the latest version, did you check that update?

Comment (francoisfreitag) [inline django/db/models/sql/compiler.py]: Oops, that's right! Thanks.

**Comment by timgraham** [comment]
Hi, did you consider how this interacts with `--keepdb`. I'm seeing errors like `cursor "_django_curs_140661412362048_20" does not exist` in some cases.

**Comment by jarshwah** [comment]
@timgraham are you getting that error fairly consistently? What test packages are you running to see this problem?

I don't see how keepdb would influence server side cursors. Were you also running tests with the parallel flag?

**Comment by timgraham** [comment]
We discussed this in private messages, I think the cause is that the original exception is hidden: https://github.com/django/django/pull/7844.
